### PR TITLE
Added 'tests' validation for scripts, playbooks and…

### PR DESCRIPTION
**Pull request by hod-alpert**
… integrations

## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
demisto/etc#23449

## Description
Validations for scripts, playbooks and integrations will now validate that `tests` key is supplied and that the test is registered in `conf.json`file unless **No tests** is supplied
## Screenshots
Paste here any images that will help the reviewer

## Must have
- [x] Tests
- [x] Documentation

### CODE REVIEW

File: demisto_sdk/commands/common/hook_validations/base_validator.py [hunk 1]
```diff
@@ -130,3 +132,71 @@ def _is_id_equals_name(self, file_type):
...
+    @staticmethod
+    def find_test_match(test_config, test_playbook_id, integration_id):
+        test_integrations = test_config.get('integrations')
+        if test_playbook_id != test_config.get('playbookID'):
+            return False
+        if isinstance(test_integrations, list):
+            return any(
+                test_integration for test_integration in test_integrations if test_integration == integration_id)
+        else:
+            return test_integrations == integration_id
```
Comment (avidan-H) [inline demisto_sdk/commands/common/hook_validations/base_validator.py:202]: I believe this logic is incorrect. As we said, test playbooks aren't only linked to integrations. With that understanding, in the `conf.json` we have test "configurations" that do not contain an `"integrations"` key... because they don't test an integration, they might be tests for a script or another playbook.

File: demisto_sdk/commands/common/hook_validations/base_validator.py [hunk 2]
```diff
@@ -130,3 +132,71 @@ def _is_id_equals_name(self, file_type):
...
+            if not test_registered_in_conf_json:
+                not_registered_tests.append(test)
+        return not_registered_tests
+
+    def yml_has_test_key(self, tests, file_type):
+        if not tests:
+            print_error(
+                f'You don\'t have a TestPlaybook for {file_type} {self.file_path}. If you don\'t want to create a'
+                f' TestPlaybook for this {file_type}, edit the yml file and add  \ntests:\n -  No tests\n lines'
+                f' to it. Or run demisto-sdk format --skip-test-playbooks -p {file_type}')
```
Comment (avidan-H) [inline demisto_sdk/commands/common/hook_validations/base_validator.py:189]: This error message isn't entirely accurate. They might have created a test playbook but it simply has not been linked to the content item. Please update the message to reflect that.

Comment (avidan-H) [reply demisto_sdk/commands/common/hook_validations/base_validator.py:189]: We shouldn't only provide the easy way out (aka by telling them about the option to add `no tests`). You need to do one of two things. 
1. **Less Ideal Option**: Before telling them how they can avoid having a test playbook, describe that they should create a test playbook and then add the ID of the test playbook to the `tests` key in the YAML of the content item - similar to how you described how they could add `No tests` themselves.
2. **More Ideal Option**: Add something like an `--attach-test-playbook` option to the `format` command where the user would give the paths to the content item and the test playbook and the command would add the ID of the test playbook to the `tests` key of the content item's YAML. Then you would describe this option first in your error message.

Given time constraints, we may have to go with the less ideal option.

File: demisto_sdk/commands/common/tests/base_validator_test.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
+import pytest
+from demisto_sdk.commands.common.hook_validations.base_validator import \
+    BaseValidator
+from demisto_sdk.commands.common.hook_validations.structure import \
+    StructureValidator
+from demisto_sdk.tests.constants_test import VALID_INTEGRATION_TEST_PATH
+
+HAS_TESTS_KEY_UNPUTS = [
+    (VALID_INTEGRATION_TEST_PATH, 'integration', True)
```
Comment (avidan-H) [inline demisto_sdk/commands/common/tests/base_validator_test.py:9]: Please also test the negative.

File: demisto_sdk/commands/common/tests/base_validator_test.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    assert validator.yml_has_test_key(tests, schema) == expected
+
+
+FIND_TEST_MATCH_IMPUT = [
+    ({'integrations': 'integration1', 'playbookID': 'playbook1'}, 'integration1', 'playbook1', True),
+    ({'integrations': 'integration1', 'playbookID': 'playbook1'}, 'integration2', 'playbook1', False),
+    ({'integrations': ['integration1', 'integration2'], 'playbookID': 'playbook1'}, 'integration1', 'playbook1', True),
+    ({'integrations': ['integration1', 'integration2'], 'playbookID': 'playbook1'}, 'integration3', 'playbook1', False),
+    ({'integrations': ['integration1', 'integration2'], 'playbookID': 'playbook1'}, 'integration1', 'playbook2', False),
+]
```
Comment (avidan-H) [inline demisto_sdk/commands/common/tests/base_validator_test.py:27]: These only test the case when the content item in question is an integration. For scripts and playbooks, only the `"playbookID"` key would necessarily be present in the test_config.

File: demisto_sdk/commands/common/tests/base_validator_test.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+        [{'integrations': 'test', 'playbookID': 'PagerDuty Test'}],
+        ['PagerDuty Test']
+    ),
+    (
+        VALID_INTEGRATION_TEST_PATH,
+        'integration',
+        [{'integrations': 'PagerDuty v2', 'playbookID': 'Playbook'}],
+        ['PagerDuty Test']
+    )
+]
```
Comment (avidan-H) [inline demisto_sdk/commands/common/tests/base_validator_test.py:54]: Same as elsewhere - need to test for playbooks and scripts as well.

File: demisto_sdk/commands/validate/file_validator.py [hunk 6]
```diff
@@ -301,15 +302,15 @@ def validate_modified_files(self, modified_files):  # noqa: C901
                 if not self.id_set_validator.is_file_valid_in_set(file_path):
                     self._is_valid = False
 
-            elif checked_type(file_path, YML_INTEGRATION_REGEXES):
+            elif checked_type(file_path, YML_INTEGRATION_REGEXES) or file_type == 'integration':
```
Comment (avidan-H) [inline demisto_sdk/commands/validate/file_validator.py:305]: What cases are you catching here (and elsewhere that you added this logic) that weren't being caught before?

File: demisto_sdk/tests/test_files/valid-beta-integration.yml [hunk 7]
```diff
@@ -46,3 +46,5 @@ script:
   type: python
   subtype: python3
 beta: true
+tests:
+  - No tests
```
Comment (avidan-H) [inline demisto_sdk/tests/test_files/valid-beta-integration.yml:50]: You will probably want to add some test files that are explicitly invalid for your unit tests.

File: demisto_sdk/commands/validate/tests/validators_test.py [hunk 8]
```diff
@@ -4,7 +4,7 @@
 from typing import Any, Type
 
 import pytest
-from demisto_sdk.commands.common.constants import DIR_LIST
+from demisto_sdk.commands.common.constants import CONF_REGEX, DIR_LIST
```
Comment (avidan-H) [inline demisto_sdk/commands/validate/tests/validators_test.py:7]: Please explain the changes you made to this file.

File: demisto_sdk/commands/common/constants.py [hunk 9]
```diff
@@ -820,6 +821,7 @@ def found_hidden_param(parameter_name):
     MISC_DIR,
     CONNECTIONS_DIR,
     INDICATOR_FIELDS_DIR,
+    TESTS_DIR
```
Comment (avidan-H) [inline demisto_sdk/commands/common/constants.py:824]: Explain this change.

File: demisto_sdk/commands/common/hook_validations/base_validator.py (hunk 2 above)
Comment (hod-alpert) [reply demisto_sdk/commands/common/hook_validations/base_validator.py:189]: The flag is not implemented currently anyways, so I'll remove the flag part until it is.

File: demisto_sdk/commands/common/hook_validations/base_validator.py [hunk 10]
```diff
@@ -130,3 +132,71 @@ def _is_id_equals_name(self, file_type):
...
+        if no_tests_explicitly:
+            return True
+        conf_json__tests = self._load_conf_file()['tests']
+
+        # Calling it integration but the validation is the same for integrations, playbooks and scripts
+        integration_id = self._get_file_id(self.structure_validator.scheme_name)
+        not_registered_tests = self.get_not_registered_tests(conf_json__tests, integration_id, tests)
+        if not_registered_tests:
+            missing_test_configurations = json.dumps([
+                {'integrations': integration_id, 'playbookID': test} for test in not_registered_tests
```
[the remark this answers is not in the document]
Comment (hod-alpert) [reply demisto_sdk/commands/common/hook_validations/base_validator.py:164]: But that's how things are going in *conf.json* file, even playbooks and scripts are written as 'integrations'.
Not cool but it is what it is.

File: demisto_sdk/commands/common/hook_validations/base_validator.py [hunk 11]
```diff
@@ -130,3 +132,71 @@ def _is_id_equals_name(self, file_type):
...
+        file_type = self.structure_validator.scheme_name
+        tests = self.current_file.get('tests', [])
+        if not self.yml_has_test_key(tests, file_type):
+            return False
+        return self.tests_registered_in_conf_json_file(tests)
+
+    def tests_registered_in_conf_json_file(self, tests):
+        # If 'No tests' is explicitly configured, we can skip checking if the tests are
+        # registered in the conf.json file.
+        no_tests_explicitly = any(test for test in tests if 'no test' in test.lower())
```
[the remark this answers is not in the document]
Comment (hod-alpert) [reply demisto_sdk/commands/common/hook_validations/base_validator.py:154]: It's a list.
When **No tests** is given it's a list with one item, but still a list.

File: demisto_sdk/commands/common/constants.py (hunk 9 above)
Comment (hod-alpert) [reply demisto_sdk/commands/common/constants.py:824]: Since base validator sometimes need to open conf.json, and conf.json is in 'Tests' folder, Added **Tests/cong.json** in the setup of test validator's test and removed it in the tearDown.

Since **Tests** folder was not created needed to import it in order to create it, makes sense?

File: demisto_sdk/commands/common/tests/base_validator_test.py (hunk 4 above)
Comment (hod-alpert) [reply demisto_sdk/commands/common/tests/base_validator_test.py:27]: For some reason all tests are appearing under integrations, even for playbooks and scripts.

File: demisto_sdk/commands/validate/file_validator.py (hunk 6 above)
Comment (hod-alpert) [reply demisto_sdk/commands/validate/file_validator.py:305]: The regexes are broken this is another way of supporting it, it's done in other methods as well until we fix the regexes.

File: demisto_sdk/tests/test_files/playbook-valid-id-test.yml [hunk 12]
```diff
@@ -455,7 +455,6 @@ tasks:
       id: 160c3d44-9b6f-4207-8df9-69f7c6ac3eb7
       version: -1
       name: IP Enrichment - Generic
-      description: ""
       playbookName: IP Enrichment - Generic
```
Comment (avidan-H) [inline demisto_sdk/tests/test_files/playbook-valid-id-test.yml:458]: How come this deletion?

File: demisto_sdk/tests/test_files/non-valid-integration-test-not-configured.yml [hunk 13]
```diff
@@ -0,0 +1,609 @@
...
+    USE_SSL = True
+
+    API_KEY = demisto.params()['APIKey']
+    SERVICE_KEY = demisto.params()['ServiceKey']
+
+    SERVER_URL = "https://api.pagerduty.com/"
+
+    DEFAULT_HEADERS = {
+                          'Authorization': 'Token token=' + API_KEY,
+                          'Accept': 'application/vnd.pagerduty+json;version=2'
```
Comment (avidan-H) [inline demisto_sdk/tests/test_files/non-valid-integration-test-not-configured.yml:40]: Do we really need all the contents of this big file if really all you're using it for is checking the `tests` field?

File: demisto_sdk/tests/test_files/non-valid-integration-no-test-playbooks.yml [hunk 14]
```diff
@@ -0,0 +1,607 @@
...
+    API_KEY = demisto.params()['APIKey']
+    SERVICE_KEY = demisto.params()['ServiceKey']
+
+    SERVER_URL = "https://api.pagerduty.com/"
+
+    DEFAULT_HEADERS = {
+                          'Authorization': 'Token token=' + API_KEY,
+                          'Accept': 'application/vnd.pagerduty+json;version=2'
+                      }
+
```
Comment (avidan-H) [inline demisto_sdk/tests/test_files/non-valid-integration-no-test-playbooks.yml:42]: Same question. Do we really need all this if all we care about is the `tests` field? Are there other validations that need all this other crap?

File: demisto_sdk/tests/test_files/integration-test-with-no-test-playbook.yml [hunk 15]
```diff
@@ -0,0 +1,607 @@
...
+    ''' GLOBAL VARS '''
+    # Pagerduty API works only with secured communication.
+    USE_SSL = True
+
+    API_KEY = demisto.params()['APIKey']
+    SERVICE_KEY = demisto.params()['ServiceKey']
+
+    SERVER_URL = "https://api.pagerduty.com/"
+
+    DEFAULT_HEADERS = {
```
Comment (avidan-H) [inline demisto_sdk/tests/test_files/integration-test-with-no-test-playbook.yml:38]: Same question as the other files.

File: demisto_sdk/tests/test_files/content_repo_example/Packs/FeedAzureValid/TestPlaybooks/script-prefixed_automation.yml [hunk 16]
```diff
@@ -15,3 +15,5 @@ args:
   description: Amount of seconds to sleep
 scripttarget: 0
 timeout: 1h40m0s
+tests:
+  - No tests
```
Comment (avidan-H) [inline demisto_sdk/tests/test_files/content_repo_example/Packs/FeedAzureValid/TestPlaybooks/script-prefixed_automation.yml]: Do other validations use these test files that you've modified? If so, please check that your additions don't create any issues in those validations.

File: demisto_sdk/tests/test_files/content_repo_example/Packs/FeedAzure/TestPlaybooks/just_a_test_script.yml [hunk 17]
```diff
@@ -15,3 +15,5 @@ args:
   description: Amount of seconds to sleep
 scripttarget: 0
 timeout: 1h40m0s
```
Comment (avidan-H) [inline demisto_sdk/tests/test_files/content_repo_example/Packs/FeedAzure/TestPlaybooks/just_a_test_script.yml:17]: Why are you using all these test files in this `FeedAzure` test pack?

File: demisto_sdk/commands/validate/tests/validators_test.py [hunk 18]
```diff
@@ -433,9 +437,20 @@ def test_script_valid_rn(self, mocker, file_path, file_type):
...
+        (INVALID_INTEGRATION_NON_CONFIGURED_TESTS, False)
+    ]
 
-def test_pack_validation():
-    file_validator = FilesValidator(validate_conf_json=False)
-    file_validator.file_path = VALID_PACK
-    file_validator.is_valid_structure()
-    assert file_validator._is_valid is False
+    @pytest.mark.parametrize('file_path, expected', ARE_TEST_CONFIGURED_TEST_INPUT)
+    def test_are_tests_configured(self, file_path, expected):
```
Comment (avidan-H) [inline demisto_sdk/commands/validate/tests/validators_test.py:437]: Please document the unit test as described [here](https://panw-global.slack.com/archives/G011E63JXPB/p1587370983267700).

File: demisto_sdk/commands/common/tests/base_validator_test.py [hunk 19]
```diff
@@ -0,0 +1,129 @@
...
+    VALID_INTEGRATION_TEST_PATH, VALID_TEST_PLAYBOOK_PATH)
+
+HAS_TESTS_KEY_UNPUTS = [
+    (VALID_INTEGRATION_TEST_PATH, 'integration', True),
+    (INVALID_INTEGRATION_WITH_NO_TEST_PLAYBOOK, 'integration', False)
+]
+
+
+@pytest.mark.parametrize('file_path, schema, expected', HAS_TESTS_KEY_UNPUTS)
+def test_yml_has_test_key(file_path, schema, expected):
```
Comment (avidan-H) [inline demisto_sdk/commands/common/tests/base_validator_test.py:17]: Please document all of your unit tests as described [here](https://panw-global.slack.com/archives/G011E63JXPB/p1587370983267700).

File: demisto_sdk/commands/common/tests/base_validator_test.py [hunk 20]
```diff
@@ -0,0 +1,129 @@
...
+    structure_validator = StructureValidator(file_path, predefined_scheme=schema)
+    validator = BaseValidator(structure_validator)
+    tests = structure_validator.current_file.get('tests')
+    assert validator.yml_has_test_key(tests, schema) == expected
+
+
+FIND_TEST_MATCH_IMPUT = [
```
Comment (avidan-H) [inline demisto_sdk/commands/common/tests/base_validator_test.py:24]:
```suggestion
FIND_TEST_MATCH_INPUT = [
```

Comment (avidan-H) [reply demisto_sdk/commands/common/tests/base_validator_test.py:24]: Fix the typo wherever else it's relevant also.

File: demisto_sdk/commands/common/tests/base_validator_test.py [hunk 21]
```diff
@@ -0,0 +1,129 @@
...
+        False
+    ),
+    (
+        {'playbookID': 'playbook1'},
+        'playbook',
+        'playbook1',
+        'playbook',
+        True
+    ),
+
```
Comment (avidan-H) [inline demisto_sdk/commands/common/tests/base_validator_test.py:67]: You have `'integration'` and `'playbook'` but not script. Please add that case as well.

File: demisto_sdk/commands/common/hook_validations/base_validator.py [hunk 22]
```diff
@@ -130,3 +132,128 @@ def _is_id_equals_name(self, file_type):
...
+            return json.load(data_file)
+
+    def are_tests_configured(self) -> bool:
+        """
+        Checks if a file (playbook or integration) has a TestPlaybook and if the TestPlaybook is configured in conf.json
+        And prints an error message accordingly
+        """
+        file_type = self.structure_validator.scheme_name
+        tests = self.current_file.get('tests', [])
+        if not self.yml_has_test_key(tests, file_type):
```
Comment (avidan-H) [inline demisto_sdk/commands/common/hook_validations/base_validator.py:147]: This treats the situation where there is no `tests` key in the YAML and the situation in which there is a `tests` key in the YAML file whose value is empty the same way. Is that intentional?

File: demisto_sdk/commands/common/hook_validations/base_validator.py [hunk 23]
```diff
@@ -130,3 +132,128 @@ def _is_id_equals_name(self, file_type):
...
+    def yml_has_test_key(self, test_playbooks: list, file_type: str) -> bool:
+        """
+        Checks if tests are configured.
+        If not: prints an error message according to the file type and return the check result
+        Args:
+            test_playbooks: The yml file's list of test playbooks
+            file_type: The file type, could be a script, an integration or a playbook.
+
+        Returns:
+            True if tests are configured (not None and not an enpty list) other-wise return False.
```
Comment (avidan-H) [inline demisto_sdk/commands/common/hook_validations/base_validator.py:219]:
```suggestion
            True if tests are configured (not None and not an empty list) otherwise return False.
```

File: demisto_sdk/commands/common/hook_validations/base_validator.py (hunk 22 above)
Comment (hod-alpert) [reply demisto_sdk/commands/common/hook_validations/base_validator.py:147]: Yes
We should either add **No tests** or add actual tests
That's the convention as I see it

File: demisto_sdk/tests/test_files/content_repo_example/Packs/FeedAzure/TestPlaybooks/just_a_test_script.yml (hunk 17 above)
Comment (hod-alpert) [reply demisto_sdk/tests/test_files/content_repo_example/Packs/FeedAzure/TestPlaybooks/just_a_test_script.yml:17]: Other tests issues

File: demisto_sdk/commands/validate/file_validator.py (hunk 6 above)
Comment (avidan-H) [reply demisto_sdk/commands/validate/file_validator.py:306]: tov

File: demisto_sdk/tests/test_files/content_repo_example/Packs/FeedAzureValid/TestPlaybooks/script-prefixed_automation.yml (hunk 16 above)
Comment (avidan-H) [reply demisto_sdk/tests/test_files/content_repo_example/Packs/FeedAzureValid/TestPlaybooks/script-prefixed_automation.yml:19]: Need to understand better why you're modifying all these test files.

Comment (hod-alpert) [reply demisto_sdk/tests/test_files/content_repo_example/Packs/FeedAzureValid/TestPlaybooks/script-prefixed_automation.yml:19]: Because the enforcement changes things for other tests too where the tests validation is irelevant.

File: CHANGELOG.md [hunk 24]
```diff
@@ -2,6 +2,7 @@
 * Added *download* command to download custom content from Demisto instance to the local content repository.
 * Improved validation failure messages to include a command suggestion, wherever relevant, to fix the raised issue.
 * Improved 'validate' help and documentation description
+* validate - checks that scripts, playbooks, and integrations have the *tests* key.
```
Comment (avidan-H) [inline CHANGELOG.md:5]: What happened to the second line?
